src/serde/de_tree.rs

```diff
@@ -0,0 +1,327 @@
+
+pub fn parse_triples<R: Read>(
+    f: &mut R,
+    calculate_tree_hashes: bool,
+) -> Result<ParsedTriplesOutput> {
+    let mut r = Vec::new();
+    let mut tree_hashes = Vec::new();
+    let mut op_stack = vec![ParseOpRef::ParseObj];
+    let mut cursor: u64 = 0;
+    loop {
+        match op_stack.pop() {
+            None => {
+                break;
+            }
+            Some(op) => match op {
+                ParseOpRef::ParseObj => {
+                    let mut b: [u8; 1] = [0];
+                    f.read_exact(&mut b)?;
+                    let start = cursor;
+                    cursor += 1;
+                    let b = b[0];
+                    if b == CONS_BOX_MARKER {
+                        let index = r.len();
+                        let new_obj = ParsedTriple::Pair {
+                            start,
+                            end: 0,
+                            right_index: 0,
+                        };
+                        r.push(new_obj);
+                        if calculate_tree_hashes {
+                            tree_hashes.push([0; 32])
+                        }
+                        op_stack.push(ParseOpRef::SaveCursor(index));
+                        op_stack.push(ParseOpRef::ParseObj);
+                        op_stack.push(ParseOpRef::SaveIndex(index));
+                        op_stack.push(ParseOpRef::ParseObj);
+                    } else {
+                        let (start, end, atom_offset, tree_hash) = {
+                            if b <= MAX_SINGLE_BYTE {
+                                (
+                                    start,
+                                    start + 1,
+                                    0,
+                                    tree_hash_for_byte(b, calculate_tree_hashes),
+                                )
+                            } else {
+                                let (atom_offset, atom_size) = decode_size_with_offset(f, b)?;
+                                let end = start + (atom_offset as u64) + atom_size;
+                                let h = skip_or_sha_bytes(f, atom_size, calculate_tree_hashes)?;
+                                (start, end, atom_offset as u32, h)
+                            }
+                        };
+                        if calculate_tree_hashes {
+                            tree_hashes.push(tree_hash.expect("failed unwrap"))
+                        }
+                        let new_obj = ParsedTriple::Atom {
+                            start,
+                            end,
+                            atom_offset,
+                        };
+                        cursor = end;
+                        r.push(new_obj);
+                    }
+                }
+                ParseOpRef::SaveCursor(index) => {
+                    if let ParsedTriple::Pair {
+                        start,
+                        end: _,
```

Comment:
can't you bind a mutable reference to `end` here, so you can just assign to it below, rather than re-creating the `Pair` object?
I think, since it would need fewer lines of code, it might be slightly easier to read.